### ISSUE TITLE
Drag doesn't work on Android

### ISSUE THREAD

**Issue by katspaugh** [katspaugh/wavesurfer.js#2936]
### Discussed in https://github.com/katspaugh/wavesurfer.js/discussions/2935

<div type='discussions-op-text'>

<sup>Originally posted by **nshen** June 25, 2023</sup>
Drag is not work on my real phone:

https://youtube.com/shorts/Cx6xMUq0B7M?feature=share</div>

**Comment by vvestman** [comment]
I would like to add couple of observations that relate to this issue:

1) Drag does not work on my touch screen Ubuntu laptop if the used browser is Edge or Chrome. Works fine on Firefox.

2) The drag selection in regions plugin does not work correctly on touch screens unless using Firefox.
I tested it with this example: [https://wavesurfer-js.org/examples/#regions.js](https://wavesurfer-js.org/examples/#regions.js)

**Comment by katspaugh** [comment]
Looks like not all browsers support the Pointer Events API. Although on caniuse, the [support is pretty good](https://caniuse.com/?search=pointer%20events).

I don't own an Android device, so I cannot debug there. On iOS Safari, it works fine.

Feel free to jump on this issue if you have an affected device. Cheers!

**Comment by vvestman** [comment]
Hi @katspaugh
I think I figured out what the problem is. In draggable.ts, the preventDefault() call in line
`const touchMove = (e: Event) => e.preventDefault()`
causes the following error on most(?) browsers:
`Ignored attempt to cancel a touchmove event with cancelable=false, for example because scrolling is in progress and cannot be interrupted.`

Basically, what ends up happening that the default scrolling behaviour is not canceled and the drag will not go on as expected.
Here is some related info:
[https://stackoverflow.com/questions/26478267/touch-move-getting-stuck-ignored-attempt-to-cancel-a-touchmove](https://stackoverflow.com/questions/26478267/touch-move-getting-stuck-ignored-attempt-to-cancel-a-touchmove)

Also, I found this old discussion:
[https://bugs.webkit.org/show_bug.cgi?id=184250](https://bugs.webkit.org/show_bug.cgi?id=184250)
And there was a comment:

> I can confirm it is only happening to event listeners that are added inside event handlers.

Based on that, I edited draggable.ts to not to add and remove listeners each time object is touched, but instead keep the listeners registered and control whether they do anything with a boolean flag. This fixed the problem on my touch screen laptop. I do not know how to work with github, but here is the proposed fix (see if you like it and find it useful):
[https://filedn.com/lgFAmApdxo3YCn6HfBSNU0R/draggable.ts](https://filedn.com/lgFAmApdxo3YCn6HfBSNU0R/draggable.ts)

**Comment by katspaugh** [comment]
`const touchMove = (e: Event) => e.preventDefault()` was added to prevent scrolling _after_ this issue had been reported.

**Comment by vvestman** [comment]
Ok! That almost solved the issue.
The problem is, however, that `e.preventDefault()` does not seem to do anything on Edge and Chrome because `e.cancelable = false` on those browsers. If `e.cancelable` would be `true`, then `e.preventDefault()` would execute properly, and it would solve this issue based on my testing. To avoid `e.cancelable = false` on these browsers, event listeners have to be added before `pointerdown` event is triggered. Currently event listeners are added after `pointerdown` event.

**Comment by vvestman** [comment]
Hello!
I tested again today with my own project and I was still having issues with dragging on Android and on Linux laptop on Chrome and Edge browsers (`e.preventDefault() `worked unpredictably; sometimes working, sometimes not). Then, I figured out that I could just put `touch-action: none;` to the css of the container containing WaveSurfer and dragging started working without problems.

I also re-tested regions example on the WaveSurfer website. Dragging in that example seems to be now working without any problems on all of my devices.
[https://wavesurfer-js.org/examples/#regions.js](https://wavesurfer-js.org/examples/#regions.js)

Thanks!
